[dynamo] Use UUID for compiled function variable names.

zhxchen17 opened the pull request:
Summary:
We previously assign each compiled function variable a name based on in-process global counter. This works fine within the same process but when we're trying to serialize the states with precompile, we need a way to load back these compiled functions without causing collision to the existing global scope.

Changing the counter to a true global uuid seems to resolve this issue.

For example, the new variable name will look like:
```
__compiled_fn_0_7ce7d872_4fe8_4174_b8fd_2496b09b8b43
```

Test Plan: CI

Differential Revision: D75244901




cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @chenyang78 @kadeng @chauhang @amjames

jansel commented:
> __compiled_fn_7ce7d872-4fe8-4174-b8fd-2496b09b8b43

Isn't this an invalid variable name?

jansel reviewed (changes requested):
Replace `-` with `_` so we get a valid identifier.

williamwen42 commented:
Can we still include the old ID as well? I find it helpful to read for bytecode/graph_code logs.

zhxchen17 committed:
[dynamo] Use UUID for compiled function variable names. (#154148)

Summary:

We previously assign each compiled function variable a name based on in-process global counter. This works fine within the same process but when we're trying to serialize the states with precompile, we need a way to load back these compiled functions without causing collision to the existing global scope.

Changing the counter to a true global uuid seems to resolve this issue.

Test Plan: CI

Differential Revision: D75244901

zhxchen17 commented:
> > __compiled_fn_7ce7d872-4fe8-4174-b8fd-2496b09b8b43
> 
> Isn't this an invalid variable name?

@jansel It seems that any valid string can be a valid python variable "name" including non-identifier https://github.com/python/steering-council/issues/142#issuecomment-1252172695

but I don't strongly favor non-identifier in this case. 

Updated to replace "-" with "_".

jansel commented:
@pytorchbot merge

ezyang commented on torch/_dynamo/bytecode_transformation.py:
```diff
@@ -1547,8 +1548,11 @@ def _cached_cleaned_instructions(code, safe=False) -> Sequence[Instruction]:
+    ret = f"{name}_{next(_unique_id_counter)}"
+    if with_uuid:
+        ret += f"_{uuid.uuid4()}".replace("-", "_")
+    return ret
```
Do you want this to be deterministic though?
